modelon/impact/client/sal/service.py

```diff
@@ -24,6 +32,36 @@ def __init__(self, uri, context=None, check_return=True):
         self.experiment = ExperimentService(self._base_uri, self._http_client)
         self.custom_function = CustomFunctionService(self._base_uri, self._http_client)
 
+        if credential_resolver is not None:
+            self._add_retry_with(credential_resolver)
+
+    def _add_retry_with(self, credential_resolver):
+        def retry_with_login_decorator(func):
+            def wrapped(*args, **kwargs):
+                try:
+                    return func(*args, **kwargs)
+                except exceptions.HTTPError as e:
+                    if e.status_code != 401:
+                        raise
+
+                    credential_resolver()
+                    return func(*args, **kwargs)
+
+            return wrapped
+
+        self.workspace = _decorate_all_methods(
+            self.workspace, retry_with_login_decorator
+        )
```

Comment:
I'm not fond of reassigning existing variables. How about having a factory-function that determines if you should decorate or not? E.g. some thing like:
self.workspace = _apply_retry_if_needed(WorkspaceService(...))
In one case it would include the credentials_resolver and decorate the methods, in the other case it would just pass the object.

Reply:
That sounds like a good idea. I will try it out.

Reply:
I ended up going in another direction. When trying to sort out so there is less implicit dependency on Client workflows in the SAL i realized that we should not add the re-try logic until we have the credentials/API key we want to it with. This leads to the re-try being added later as a separate call to the SAL where everything is much more explicit (there is no callback that in itself would call the SAL...).